plugin/manager.go

```diff
@@ -0,0 +1,277 @@
+}
+
+// Notifier notifies when a new message was created.
+type Notifier interface {
+	Notify(userID uint, message *model.Message)
+}
+
+type Manager struct {
+	mutex     *sync.RWMutex
+	instances map[uint]compat.PluginInstance
+	plugins   map[string]compat.Plugin
+	messages  chan MessageWithUserID
+	db        PluginDatabase
+	mux       *gin.RouterGroup
+}
+
+func NewManager(db PluginDatabase, directory string, mux *gin.RouterGroup, notifier Notifier) (*Manager, error) {
+	manager := &Manager{
+		mutex:     &sync.RWMutex{},
+		instances: map[uint]compat.PluginInstance{},
+		plugins:   map[string]compat.Plugin{},
+		messages:  make(chan MessageWithUserID),
+		db:        db,
+		mux:       mux,
+	}
+
+	if err := manager.loadPlugins(directory); err != nil {
+		return nil, err
+	}
+
+	for _, user := range manager.db.GetUsers() {
+		if err := manager.initializeForUser(*user); err != nil {
+			return nil, err
+		}
+	}
+
+	go func() {
+		for {
+			message := <-manager.messages
+			db.CreateMessage(&message.Message)
+			notifier.Notify(message.UserID, &message.Message)
+		}
+	}()
+
+	return manager, nil
+}
+
+func (m *Manager) SetPluginEnabled(pluginID uint, enabled bool) error {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	instance, err := m.Instance(pluginID)
+	if err != nil {
+		return errors.New("instance not found")
+	}
+	conf := m.db.GetPluginConfByID(pluginID)
+
+	if conf.Enabled == enabled {
+		return errors.New("config is already enabled/disabled")
+	}
+
+	if enabled {
+		err = instance.Enable()
+	} else {
+		err = instance.Disable()
+	}
+	if err != nil {
+		return err
+	}
+	conf.Enabled = enabled
+	return m.db.UpdatePluginConf(conf)
+}
+
+func (m *Manager) PluginInfo(modulePath string) compat.Info {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	if p, ok := m.plugins[modulePath]; ok {
+		return p.PluginInfo()
+	}
+	fmt.Println("Could not get plugin info for", modulePath)
+	return compat.Info{
+		Name:        "UNKNOWN",
+		ModulePath:  modulePath,
+		Description: "Oops something went wrong",
+	}
+}
+
+func (m *Manager) Instance(pluginID uint) (compat.PluginInstance, error) {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	if instance, ok := m.instances[pluginID]; ok {
+		return instance, nil
+	}
+	return nil, errors.New("instance not found")
+}
+
+func (m *Manager) HasInstance(pluginID uint) bool {
+	instance, err := m.Instance(pluginID)
+	return err == nil && instance != nil
+}
+
+func (m *Manager) RemoveUser(userID uint) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	for _, p := range m.plugins {
+		pluginConf := m.db.GetPluginConfByUserAndPath(userID, p.PluginInfo().ModulePath)
+		if pluginConf.Enabled {
+			inst, err := m.Instance(pluginConf.ID)
+			if err != nil {
+				return err
+			}
+			if err := inst.Disable(); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (m *Manager) loadPlugins(directory string) error {
+	if directory == "" {
+		return nil
+	}
+
+	if err := os.MkdirAll(directory, 0777); err != nil {
```

Comment:
Firstly, just don't create directories with 0777 permission. Use 0755

Secondly, maybe consider moving the directory creation proc to `app.go`, because the image dir is created there(also use 0644 or 0666 for the image dir), or move them both to an appropriate location.

Reply:
Yeah, true. Feel free to move it (:.